src/resmom/requests.c

```diff
@@ -1622,12 +1628,171 @@ post_resume(job *pjob, int err)
 		pjob->ji_flags &= ~MOM_SISTER_ERR;
 }
 
+#if MOM_ALPS 
+
+/*
+ * Try to minimize latency of suspend/resume. Wait half a second before
+ * the first check, and then poll every tenth of a second.
+ */
+
+#define ALPS_SWITCH_SLEEP_USECS_LONG (500000)
+#define ALPS_SWITCH_SLEEP_USECS_SHORT (100000)
+
+/**
+ * On a Cray, make the requested switch, and confirm it 
+ * @param[in]	pjob	job of interest
+ * @param[in]	which	SUSPEND/RESUME
+ * @retval	PBSE_NONE	no error
+ * @retval	PBSE_ALPS_SWITCH_ERR
+ */
+static int
+do_cray_susres_conf (job *pjob, int which)
+{ 
+	/**
+	 * On a Cray, we need to send an ALPS SWITCH request to move the jobs
+	 * to a suspend or resume state.
+	 */
+	basil_switch_action_t action;
+	int i;
+	int rc = 0;
+	time_t	total_time = 0;
+	time_t	begin_time = 0;
+	time_t	end_time = 0;
+	int	timeout_val = alps_confirm_switch_timeout;
+	int	first_status;
+	int	first_status_was_empty = 0;
+	int	first_sleep;
+	
+	/* Check if there is an ALPS reservation to act on.
+	 * If not, just return PBSE_NONE.  
+	 */
+	if (pjob->ji_extended.ji_ext.ji_reservation <= 0) {
+		return PBSE_NONE;
```

Comment:
I think this condition is added just for safety when a job may not report alps reservation. This to me is unlikely. I also think that switch error are more suited for cases where we tried a switch and failed, in this case we can not even try a switch because there is no reservation.
what do you think?

Reply:
I suppose it depends on how the caller(s) handle the return code. If they don't care whether the reservation exists, then this is as good a way as any to handle the situation. At the very least, we should be printing a log message when this condition occurs... probably an INFO or WARN.

Reply:
This is handling the case where a job running on a login node needs to be suspended, in which case there is no need to talk to ALPS to do a SWITCH.

Reply:
Not exactly. This is handling a case where the reservation doesn't exist on the mom. Still the case that ALPS doesn't have to perform a switch, but mom expected the reservation to be there. We should log something when this happens.

Reply:
I was wrong here, this code is not only for safety but there is another use case where it will come in handy - when the job runs on login node and that job is being preempted. This is an unlikely scenario but it can still happen. In those cases we might to just continue as we would on any non cray platform and then suspend the job. So, I'm also of the opinion to not log a message. What do you think @mike0042 ?